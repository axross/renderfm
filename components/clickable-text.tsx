import React from "react";
import { twMerge } from "tailwind-merge";
import { primaryFocusRing } from "~/constants/style";

const clickableTextCss = twMerge(
  primaryFocusRing,
  "rounded-sm text-yellow-600 dark:text-yellow-400 hover:underline"
);

export interface ClickableTextProps {
  onClick?: React.MouseEventHandler<HTMLSpanElement>;
  className?: string;
  style?: React.CSSProperties;
  children?: React.ReactNode;
}

export const ClickableText: React.VFC<ClickableTextProps> = ({
  className,
  children,
  ...props
}) => {
  return (
    <span
      className={twMerge(clickableTextCss, "cursor-pointer", className)}
      tabIndex={0}
      {...props}
    >
      {children}
    </span>
  );
};

export type TextAnchorProps = React.AnchorHTMLAttributes<HTMLAnchorElement>;

export const TextAnchor: React.VFC<TextAnchorProps> = ({
  href,
  className,
  children,
  ...props
}) => {
  return (
    <a href={href} className={twMerge(clickableTextCss, className)} {...props}>
      {children}
    </a>
  );
};
