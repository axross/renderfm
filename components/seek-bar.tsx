import * as React from "react";
import { twMerge } from "tailwind-merge";
import { useRect } from "~/hooks/use-rect";

export interface SeekBarProps {
  value?: number;
  defaultValue?: number;
  disabled?: boolean;
  onChange?: (value: number) => void;
  className?: string;
  style?: React.CSSProperties;
}

export const SeekBar: React.VFC<SeekBarProps> = ({
  value: givenValue,
  defaultValue = 0,
  disabled = false,
  onChange = () => undefined,
  className,
  ...props
}) => {
  if (typeof givenValue === "number" && (givenValue < 0 || givenValue > 1)) {
    throw new Error("props.value needs to be 0 <= value <= 1.");
  }

  const [ref, rect] = useRect<HTMLDivElement>();
  const [value, setValue] = React.useState(givenValue ?? defaultValue);
  const [isDragging, setDragging] = React.useState(false);
  const [draggingX, setDraggingX] = React.useState(0);
  const _value = givenValue ?? value;

  const onMouseDown = (e: React.MouseEvent<HTMLElement>) => {
    if (disabled) {
      return;
    }

    const mouseMoveEventListner = (e: MouseEvent) => {
      const relativeX = Math.min(
        Math.max(e.clientX - rect.left, 0),
        rect.width
      );
      const v = relativeX / rect.width;

      if (givenValue === undefined) {
        setDraggingX(relativeX);
        setValue(v);
      }

      onChange(v);
    };

    const mouseUpEventListener = () => {
      window.removeEventListener("mousemove", mouseMoveEventListner);
      window.removeEventListener("mouseup", mouseUpEventListener);

      setDragging(false);
    };

    window.addEventListener("mousemove", mouseMoveEventListner);
    window.addEventListener("mouseup", mouseUpEventListener);

    const relativeX = Math.min(Math.max(e.clientX - rect.left, 0), rect.width);
    const v = relativeX / rect.width;

    setDragging(true);

    if (givenValue === undefined) {
      setDraggingX(relativeX);
      setValue(v);
    }

    onChange(v);
  };

  return (
    <div
      className={twMerge(
        "relative h-2 bg-slate-500/50 rounded-full cursor-pointer",
        disabled ? "cursor-not-allowed" : null,
        className
      )}
      onMouseDown={onMouseDown}
      role="meter"
      aria-valuemin={0}
      aria-valuemax={1}
      aria-valuenow={_value}
      ref={ref}
      {...props}
    >
      <div
        className="absolute top-0 left-0 bottom-0 bg-yellow-500 disabled:bg-yellow-500/50 rounded-full"
        style={{
          width:
            isDragging && givenValue === undefined
              ? `${draggingX}px`
              : `${_value * rect.width}px`,
          transitionProperty: isDragging ? "none" : undefined,
        }}
        data-testid="indicator-active"
      />
    </div>
  );
};
