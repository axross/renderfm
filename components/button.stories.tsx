import * as React from "react";
import { ComponentStory, ComponentMeta } from "@storybook/react";
import { Button } from "~/components/button";

export default {
  title: "Components/Button",
  component: Button,
  args: {
    variant: "primary",
    disabled: false,
    children: "Press Me",
  },
} as ComponentMeta<typeof Button>;

export const Example: ComponentStory<typeof Button> = (args) => (
  <Button {...args} />
);

export const Small = Example.bind({});
Small.args = {
  size: "sm",
};

export const ExtraSmall = Example.bind({});
ExtraSmall.args = {
  size: "xs",
};

export const Translucent = Example.bind({});
Translucent.args = {
  translucent: true,
};

export const Disabled = Example.bind({});
Disabled.args = {
  disabled: true,
};

export const Loading = Example.bind({});
Loading.args = {
  loading: true,
};
