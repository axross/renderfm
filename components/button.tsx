import { twMerge } from "tailwind-merge";
import * as React from "react";
import { LoadingIcon } from "~/components/icon";
import { primaryFocusRing } from "~/constants/style";

export type ButtonVariant = "primary";
export type ButtonSize = "base" | "sm" | "xs";

const DEFAULT_VARIANT: ButtonVariant = "primary";
const DEFAULT_SIZE: ButtonSize = "base";

export interface ButtonProps
  extends React.ButtonHTMLAttributes<HTMLButtonElement> {
  variant?: ButtonVariant;
  size?: ButtonSize;
  translucent?: boolean;
  loading?: boolean;
}

export const Button: React.VFC<ButtonProps> = ({
  variant = DEFAULT_VARIANT,
  size = DEFAULT_SIZE,
  loading = false,
  translucent = false,
  disabled,
  className,
  children,
  ...props
}) => {
  return (
    <button
      disabled={disabled || loading}
      className={twMerge(
        "inline-flex items-center rounded-md disabled:cursor-not-allowed transition-all",
        variant === "primary"
          ? translucent
            ? "bg-yellow-600/20 hover:bg-yellow-600/30 disabled:bg-yellow-600/10 text-yellow-600"
            : "bg-yellow-600 disabled:bg-yellow-600/50 text-white"
          : null,
        variant === "primary" ? primaryFocusRing : null,
        size === "base" ? "px-4 py-2 text-base" : null,
        size === "sm" ? "px-3 py-1.5 text-base" : null,
        size === "xs" ? "px-2 py-1 text-sm" : null,
        className
      )}
      {...props}
    >
      {loading ? (
        <LoadingIcon
          className={twMerge(
            size === "base" ? "w-5 h-5 -ml-1.5 mr-2" : null,
            size === "sm" ? "w-4 h-4 -ml-1 mr-1.5" : null,
            size === "xs" ? "w-3 h-3 mr-1" : null
          )}
        />
      ) : null}

      {children}
    </button>
  );
};
