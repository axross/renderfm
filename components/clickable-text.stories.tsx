import { ComponentStory, ComponentMeta } from "@storybook/react";
import { ClickableText } from "~/components/clickable-text";

export default {
  title: "Components/ClickableText",
  component: ClickableText,
  args: {
    children: "This is a link",
  },
} as ComponentMeta<typeof ClickableText>;

export const Example: ComponentStory<typeof ClickableText> = (args) => (
  <ClickableText {...args} />
);
