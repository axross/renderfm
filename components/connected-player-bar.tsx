import { twMerge } from "tailwind-merge";
import { PlayerBar } from "~/components/player-bar";
import { useHotkeyEffect } from "~/hooks/use-hotkey-effect";
import { useAudioPlayer } from "use-audio-player";
import React from "react";

interface ConnectedPlayerBar {
  className?: string;
  style?: React.CSSProperties;
}

export const ConnectedPlayerBar: React.VFC<ConnectedPlayerBar> = ({
  className,
  ...props
}) => {
  const player = useAudioPlayer();

  useHotkeyEffect(
    "ArrowLeft",
    () => {
      if (!player.isPlayable) return;

      player.seek(-15);
    },
    [player]
  );

  useHotkeyEffect(
    "ArrowRight",
    () => {
      if (!player.isPlayable) return;

      player.seek(15);
    },
    [player]
  );

  useHotkeyEffect(
    " ",
    () => {
      if (!player.isPlayable) return;

      if (player.isPlaying) {
        player.pause();
      } else {
        player.play();
      }
    },
    [player]
  );

  return (
    <PlayerBar
      trackTitle={player.track?.title ?? ""}
      duration={player.duration}
      currentTime={player.currentTime}
      playing={player.isPlaying}
      durationLoaded={player.isDurationLoaded}
      playbackLoading={player.isPlaybackLoading}
      onPauseButtonClick={player.pause}
      onResumeButtonClick={player.play}
      onRewindButtonClick={React.useCallback(() => player.seek(-15), [player])}
      onForwardButtonClick={React.useCallback(() => player.seek(15), [player])}
      onSeek={(d) => player.seekTo(player.duration * d)}
      onCloseButtonClick={player.unload}
      className={twMerge(
        "fixed left-0 right-0 transition-all",
        player.track !== null
          ? "bottom-0"
          : "-bottom-40 pointer-fine:-bottom-24",
        className
      )}
      {...props}
    />
  );
};
