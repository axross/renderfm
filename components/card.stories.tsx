import * as React from "react";
import { ComponentStory, ComponentMeta } from "@storybook/react";
import { Button } from "~/components/button";
import { Card, CardFooter } from "~/components/card";

export default {
  title: "Components/Card",
  component: Card,
  args: {
    className: "w-96 min-h-64",
    children: "Hello",
  },
} as ComponentMeta<typeof Card>;

export const Example: ComponentStory<typeof Card> = (args) => (
  <Card {...args} />
);

export const WithFooter = Example.bind({});
WithFooter.args = {
  children: (
    <>
      Hello
      <CardFooter>
        <Button>Do Some Action</Button>
      </CardFooter>
    </>
  ),
};
