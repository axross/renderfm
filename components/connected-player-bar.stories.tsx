import * as React from "react";
import { ComponentStory, ComponentMeta } from "@storybook/react";
import { Button } from "~/components/button";
import { ConnectedPlayerBar } from "~/components/connected-player-bar";
import {
  AudioPlayerProvider,
  createBrowserAudioPlayer,
  useAudioPlayer,
} from "use-audio-player";

export default {
  title: "Connected Components/PlayerBar",
  component: ConnectedPlayerBar,
  decorators: [
    (Story) => {
      const player = createBrowserAudioPlayer();

      return (
        <AudioPlayerProvider player={player}>
          <Story />
        </AudioPlayerProvider>
      );
    },
  ],
} as ComponentMeta<typeof ConnectedPlayerBar>;

export const Example: ComponentStory<typeof ConnectedPlayerBar> = (args) => {
  const player = useAudioPlayer();

  return (
    <>
      <Button
        onClick={() => {
          player.load({
            title: "Some Fancy Music",
            src: "https://file-examples-com.github.io/uploads/2017/11/file_example_MP3_5MG.mp3",
          });

          // eslint-disable-next-line storybook/await-interactions, storybook/context-in-play-function
          player.play();
        }}
      >
        Play Something
      </Button>

      <ConnectedPlayerBar {...args} />
    </>
  );
};
