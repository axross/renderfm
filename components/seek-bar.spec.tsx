jest.mock("~/hooks/use-rect", () => {
  const React = jest.requireActual<typeof import("react")>("react");

  return {
    useRect: () => {
      const ref = React.useRef<HTMLElement>(null);

      return [
        ref,
        {
          top: 12,
          bottom: 34,
          left: 56,
          right: 67,
          width: 890,
          height: 123,
        },
      ];
    },
  };
});

import { fireEvent, render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { SeekBar, SeekBarProps } from "~/components/seek-bar";

describe("<SeekBar>", () => {
  function setup(additionalProps?: Partial<SeekBarProps>) {
    const onChange = jest.fn();

    const { rerender } = render(
      <SeekBar
        onChange={onChange}
        data-testid="seek-bar"
        {...additionalProps}
      />
    );

    const seekBar = () => screen.getByTestId("seek-bar");

    return {
      seekBar,
      onChange,
      rerender: (newProps?: Partial<SeekBarProps>) =>
        rerender(
          <SeekBar data-testid="seek-bar" {...additionalProps} {...newProps} />
        ),
    };
  }

  const setupDisabled: typeof setup = (additionalProps) =>
    setup({ disabled: true, ...additionalProps });

  const setupWithDefaultValue: typeof setup = (additionalProps) =>
    setup({ defaultValue: 0.5, ...additionalProps });

  it("renders an uncontrolled seek bar", () => {
    const { seekBar } = setup();

    expect(seekBar()).toMatchSnapshot();
  });

  it("renders an uncontrolled seek bar with the given defaultValue", () => {
    const { seekBar } = setupWithDefaultValue();

    expect(seekBar()).toMatchSnapshot();
  });

  it("calls onChange() with updating the internal value whenever somewhere on the seek bar gets clicked", () => {
    const { seekBar, onChange } = setup();

    userEvent.click(seekBar(), { clientX: 234 });

    expect(seekBar()).toHaveAttribute("aria-valuenow", "0.2");

    userEvent.click(seekBar(), { clientX: 501 });

    expect(seekBar()).toHaveAttribute("aria-valuenow", "0.5");

    expect(onChange).toHaveBeenCalledTimes(2);
    expect(onChange).toHaveBeenNthCalledWith(1, 0.2);
    expect(onChange).toHaveBeenNthCalledWith(2, 0.5);
  });

  it("calls onChange() with updating the internal value whenever it's drag-n-dropped the indicator edge", () => {
    const { seekBar, onChange } = setup();

    fireEvent.mouseDown(seekBar(), { clientX: 501 });

    expect(seekBar()).toHaveAttribute("aria-valuenow", "0.5");

    fireEvent.mouseMove(seekBar(), { clientX: 408 });

    expect(seekBar()).toHaveAttribute("aria-valuenow", "0.3955056179775281");

    fireEvent.mouseMove(seekBar(), { clientX: 320 });

    expect(seekBar()).toHaveAttribute("aria-valuenow", "0.2966292134831461");

    fireEvent.mouseMove(seekBar(), { clientX: 234 });

    expect(seekBar()).toHaveAttribute("aria-valuenow", "0.2");

    fireEvent.mouseMove(seekBar(), { clientX: 56 });

    expect(seekBar()).toHaveAttribute("aria-valuenow", "0");

    fireEvent.mouseUp(seekBar(), { clientX: 56 });

    expect(onChange).toHaveBeenCalledTimes(5);
    expect(onChange).toHaveBeenNthCalledWith(1, 0.5);
    expect(onChange).toHaveBeenNthCalledWith(2, 0.3955056179775281);
    expect(onChange).toHaveBeenNthCalledWith(3, 0.2966292134831461);
    expect(onChange).toHaveBeenNthCalledWith(4, 0.2);
    expect(onChange).toHaveBeenNthCalledWith(5, 0);
  });

  it("doesn't update the internal value when disabled=true is given", () => {
    const { seekBar } = setupDisabled();

    userEvent.click(seekBar(), { clientX: 234 });

    expect(seekBar()).toHaveAttribute("aria-valuenow", "0");

    userEvent.click(seekBar(), { clientX: 501 });

    expect(seekBar()).toHaveAttribute("aria-valuenow", "0");

    fireEvent.mouseDown(seekBar(), { clientX: 501 });

    expect(seekBar()).toHaveAttribute("aria-valuenow", "0");

    fireEvent.mouseMove(seekBar(), { clientX: 408 });

    expect(seekBar()).toHaveAttribute("aria-valuenow", "0");

    fireEvent.mouseMove(seekBar(), { clientX: 320 });

    expect(seekBar()).toHaveAttribute("aria-valuenow", "0");

    fireEvent.mouseMove(seekBar(), { clientX: 234 });

    expect(seekBar()).toHaveAttribute("aria-valuenow", "0");

    fireEvent.mouseMove(seekBar(), { clientX: 120 });

    expect(seekBar()).toHaveAttribute("aria-valuenow", "0");

    fireEvent.mouseUp(seekBar(), { clientX: 120 });
  });

  it("doesn't call onChange() when disabled=true is given", () => {
    const { seekBar, onChange } = setupDisabled();

    userEvent.click(seekBar(), { clientX: 234 });
    userEvent.click(seekBar(), { clientX: 501 });
    fireEvent.mouseDown(seekBar(), { clientX: 501 });
    fireEvent.mouseMove(seekBar(), { clientX: 408 });
    fireEvent.mouseMove(seekBar(), { clientX: 320 });
    fireEvent.mouseMove(seekBar(), { clientX: 234 });
    fireEvent.mouseMove(seekBar(), { clientX: 56 });
    fireEvent.mouseUp(seekBar(), { clientX: 56 });

    expect(onChange).not.toHaveBeenCalled();
  });

  it("renders a controlled seek bar when props.value is given", () => {
    const { seekBar, rerender } = setup({ value: 0.5 });

    expect(seekBar()).toMatchSnapshot();

    rerender({ value: 0.762 });

    expect(seekBar()).toMatchSnapshot();
  });
});
