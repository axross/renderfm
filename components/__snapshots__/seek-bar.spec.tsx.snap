// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SeekBar> renders a controlled seek bar when props.value is given 1`] = `
<div
  aria-valuemax="1"
  aria-valuemin="0"
  aria-valuenow="0.5"
  class="relative h-2 bg-slate-500/50 rounded-full cursor-pointer"
  data-testid="seek-bar"
  role="meter"
>
  <div
    class="absolute top-0 left-0 bottom-0 bg-yellow-500 disabled:bg-yellow-500/50 rounded-full"
    data-testid="indicator-active"
    style="width: 445px;"
  />
</div>
`;

exports[`<SeekBar> renders a controlled seek bar when props.value is given 2`] = `
<div
  aria-valuemax="1"
  aria-valuemin="0"
  aria-valuenow="0.762"
  class="relative h-2 bg-slate-500/50 rounded-full cursor-pointer"
  data-testid="seek-bar"
  role="meter"
>
  <div
    class="absolute top-0 left-0 bottom-0 bg-yellow-500 disabled:bg-yellow-500/50 rounded-full"
    data-testid="indicator-active"
    style="width: 678.1800000000001px;"
  />
</div>
`;

exports[`<SeekBar> renders an uncontrolled seek bar 1`] = `
<div
  aria-valuemax="1"
  aria-valuemin="0"
  aria-valuenow="0"
  class="relative h-2 bg-slate-500/50 rounded-full cursor-pointer"
  data-testid="seek-bar"
  role="meter"
>
  <div
    class="absolute top-0 left-0 bottom-0 bg-yellow-500 disabled:bg-yellow-500/50 rounded-full"
    data-testid="indicator-active"
    style="width: 0px;"
  />
</div>
`;

exports[`<SeekBar> renders an uncontrolled seek bar with the given defaultValue 1`] = `
<div
  aria-valuemax="1"
  aria-valuemin="0"
  aria-valuenow="0.5"
  class="relative h-2 bg-slate-500/50 rounded-full cursor-pointer"
  data-testid="seek-bar"
  role="meter"
>
  <div
    class="absolute top-0 left-0 bottom-0 bg-yellow-500 disabled:bg-yellow-500/50 rounded-full"
    data-testid="indicator-active"
    style="width: 445px;"
  />
</div>
`;
